lib/Sema/TypeCheckAttr.cpp

```diff
@@ -2260,6 +2260,13 @@ void AttributeChecker::checkApplicationMainAttribute(DeclAttribute *attr,
     attr->setInvalid();
   }
 
+  diagnose(attr->getLocation(),
+           diag::attr_ApplicationMain_deprecated,
+           applicationMainKind)
+    .warnUntilSwiftVersion(6)
```

Comment:
`warnUntilSwiftVersion(6)` will wrap the given error message into `diag::error_in_future_swift_version` when produced as a warning. So the warning message produced in Swift 5 mode will be `<<original message>>; this is an error in Swift 6`, and the error produced in Swift 6 is `<<original message>>`.

I recommend removing the "and will be removed in a future release" part of the message, because the important information for the programmer is that this will become an error in Swift 6. Perhaps also consider shuffling around the message to avoid two `;`s.

Reply:
I've split it into a primary diagnostic which we can attach this trailing cause to and a note.